src/native/eventpipe/ds-dump-protocol.c

```diff
@@ -93,6 +93,101 @@ dump_protocol_helper_unknown_command (
 	return true;
 }
 
+static
+void
+dump_protocol_generate_core_dump_response_init(
+	DiagnosticsGenerateCoreDumpResponsePayload *payload,
+	ds_ipc_result_t error,
+	ep_char8_t *errorText)
+{
+	EP_ASSERT (payload != NULL);
+
+	payload->error = error;
+	payload->error_message = ep_rt_utf8_to_utf16_string (errorText, -1);
+}
+
+static
+void
+dump_protocol_generate_core_dump_response_fini(
+	DiagnosticsGenerateCoreDumpResponsePayload *payload)
+{
+	ep_rt_utf16_string_free (payload->error_message);
+}
+
+static
+uint16_t
+dump_protocol_generate_core_dump_response_get_size (
+	DiagnosticsGenerateCoreDumpResponsePayload *payload)
+{
+	EP_ASSERT (payload != NULL);
+
+	size_t size = sizeof(payload->error);
+
+	size += sizeof(uint32_t);
+	size += (payload->error_message != NULL) ? (ep_rt_utf16_string_len (payload->error_message) + 1) * sizeof(ep_char16_t) : 0;
+
+	EP_ASSERT (size <= UINT16_MAX);
+	return (uint16_t)size;
+}
+
+static
+bool
+dump_protocol_generate_core_dump_response_flatten (
+	void *payload,
+	uint8_t **buffer,
+	uint16_t *size)
+{
+	DiagnosticsGenerateCoreDumpResponsePayload *response = (DiagnosticsGenerateCoreDumpResponsePayload*)payload;
+
+	EP_ASSERT (payload != NULL);
+	EP_ASSERT (buffer != NULL);
+	EP_ASSERT (*buffer != NULL);
+	EP_ASSERT (size != NULL);
+	EP_ASSERT (dump_protocol_generate_core_dump_response_get_size (response) == *size);
+
+	bool success = true;
+
+	// int32_t error
+	memcpy (*buffer, &response->error, sizeof (response->error));
+	*buffer += sizeof (response->error);
+	*size -= sizeof (response->error);
+
+	// LPCWSTR error message
+	success &= ds_ipc_message_try_write_string_utf16_t (buffer, size, response->error_message);
```

Comment:
Should this handle case if response->error_message is NULL due to allocation failure in init, alternative harden ds_ipc_message_try_write_string_utf16_t to handle case where value is NULL instead of just doing an EP_ASSERT that value is not NULL?